Sources/Tools/SoundDecompress/main.cpp

```diff
@@ -0,0 +1,258 @@
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+// IMPORTANT!
+// If you want to use this to generate files compatible with older clients that
+// expect cached sounds in WAV format, you *must* set this to 1.
+#define OUTPUT_WAV_FILES 0
+
+#include <utility>
+
+#include "plResMgr/plResManager.h"
+#include "plResMgr/plResMgrSettings.h"
+#include "plResMgr/plRegistryHelpers.h"
+#include "plResMgr/plRegistryNode.h"
+
+#include "plAgeDescription/plAgeManifest.h"
+
+#include "plAudioCore/plSoundBuffer.h"
+
+#if OUTPUT_WAV_FILES
+#include "plOldAudioFileReader.h"
+#endif
+
+typedef std::set<std::pair<ST::string, uint16_t>> SoundSet;
+
+enum OutputStyle
+{
+    kSilent,
+    kProgress,
+    kVerbose
+};
+
+//// plSoundBufferCollector //////////////////////////////////////////////////
+//  Page iterator that collects all the plSoundBuffers in all of our pages
+
+class plSoundBufferCollector : public plRegistryPageIterator, public plKeyCollector
+{
+public:
+    plSoundBufferCollector(std::set<plKey>& keyArray)
+                : plKeyCollector(keyArray) {}
+
+    bool EatPage(plRegistryPageNode* page)
+    {
+        if (page->IsValid())
+        {
+            page->LoadKeys();
+            return page->IterateKeys(this, plSoundBuffer::Index());
+        }
+        else
+        {
+            fprintf(stderr, "INVALID PAGE: %s\n", page->GetPagePath().AsString().c_str());
+            return true;
+        }
+    }
+};
+
+
+void PrintHelp()
+{
+    printf("Plasma Sound Decompressor\n");
+    printf("-------------------------\n");
+    printf("-s, --silent        Run silently, no output\n");
+    printf("-v, --verbose       Print each filename when decompressing\n");
+    printf("-f, --force         Force decompressing existing files\n");
+}
+
+bool CollectSounds(plResManager* rm, SoundSet& sfxArray)
```

Comment:
It might be better to return just the set. NRVO will make sure we don't make unnecessary copies.